api/v1beta1/disruption_types.go

```diff
@@ -91,13 +94,85 @@ func init() {
 }
 
 // Hash returns the disruption spec JSON hash
-func (ds *DisruptionSpec) Hash() (string, error) {
+func (s *DisruptionSpec) Hash() (string, error) {
 	// serialize instance spec to JSON
-	specBytes, err := json.Marshal(ds)
+	specBytes, err := json.Marshal(s)
 	if err != nil {
 		return "", fmt.Errorf("error serializing instance spec: %w", err)
 	}
 
 	// compute bytes hash
 	return fmt.Sprintf("%x", md5.Sum(specBytes)), nil
 }
+
+func (s *DisruptionSpec) Validate() error {
+	// Rule: no targeted container if disruption is node-level
+	if len(s.Containers) > 0 && s.Level != chaostypes.DisruptionLevelPod {
+		return errors.New("cannot target specific container of a node-level disruption")
+	}
+
+	// Rule: at least one disruption field
+	if s.DNS == nil &&
+		s.CPUPressure == nil &&
+		s.Network == nil &&
+		s.NodeFailure == nil &&
+		s.DiskPressure == nil {
+		return errors.New("cannot apply a disruption with no target process")
+	}
+
+	//Rule: DNS and Network disruptions are incompatible
+	if s.DNS != nil && s.Network != nil {
+		return errors.New("cannot apply DNS and Network disruptions concurrently")
+	}
+
+	return nil
+}
+
+// validates rules for disruption global scope and all subsequent disruption specifications
+func (s *DisruptionSpec) ValidateDisruptionSpec() error {
+	err := s.Validate()
+	if err != nil {
+		return err
+	}
+
+	for _, kind := range chaostypes.DisruptionKinds {
+		var validator chaosapi.DisruptionValidator
+
+		disruptionExists, validator, _ := s.DisruptionKindInterfaceGenerator(kind)
+		if !disruptionExists {
+			continue
+		}
+
+		err := validator.Validate()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// validates existence and generates instances of DisruptionKind Interfaces: DisruptionValidator, DisruptionArgsGenerator
+func (s *DisruptionSpec) DisruptionKindInterfaceGenerator(kind chaostypes.DisruptionKind) (bool, chaosapi.DisruptionValidator, chaosapi.DisruptionArgsGenerator) {
```

Comment:
I've checked all the call-sites of this method, and you can simplify the return values here. Ideally, we could return only a single value, which is one of `s.NodeFailure`, `s.Network`, `s.DNS`, `s.CPUPressure`, or `s.DiskPressure`. The `bool` isn't needed, because the call-site can just use `!= nil` as they did before. We wouldn't have to return the two interfaces separately, because the returned value actually satisfies both of them. 

Why don't we do this currently? Well it's because that there's no explicit relationship between the two interface types, and there's no other "interface" or "type" that the five kinds satisfy. However, clearly from how you've been forced to return the same object twice in this fashion, it seems to me that there should be! I recommend collapsing `chaosapi.DisruptionValidator` and `chaosapi.DisruptionArgsGenerator` into a single interface that all five kinds satisfy, and making that be the sole return value here.

Then this method becomes purely a "map the name of a kind into the real kind" function, which is what the call-sites all really want.